hip/base/kernel_launch.hip.hpp

```diff
@@ -0,0 +1,260 @@
+compact_dense_wrapper<const hip_type<ValueType>> compact(
+    const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {as_hip_type(mtx->get_const_values())};
+}
+
+
+template <typename ValueType>
+hip_type<ValueType> *vector(matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return as_hip_type(mtx->get_values());
+}
+
+
+template <typename ValueType>
+const hip_type<ValueType> *vector(const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return as_hip_type(mtx->get_const_values());
+}
+
+
+template <typename T>
+struct device_unpack_1d_impl {
+    using type = T;
+    static __device__ __forceinline__ type unpack(T param, size_type, size_type)
+    {
+        return param;
+    }
+};
+
+
+template <typename T>
+struct device_unpack_2d_impl {
+    using type = T;
+    static __device__ __forceinline__ type unpack(T param, size_type, size_type,
+                                                  size_type, size_type)
+    {
+        return param;
+    }
+};
+
+template <typename ValueType>
+struct device_unpack_2d_impl<compact_dense_wrapper<ValueType>> {
+    using type = matrix_accessor<ValueType>;
+    static __device__ __forceinline__ type
+    unpack(compact_dense_wrapper<ValueType> param, size_type, size_type,
+           size_type, size_type num_cols)
+    {
+        return {param.data, num_cols};
+    }
+};
+
+
+template <typename KernelFunction, typename... KernelArgs>
+__global__ __launch_bounds__(default_block_size) void generic_kernel_1d(
+    size_type size, KernelFunction fn, KernelArgs... args)
+{
+    auto tidx = thread::get_thread_id_flat();
+    if (tidx >= size) {
+        return;
+    }
+    fn(tidx, device_unpack_1d_impl<KernelArgs>::unpack(args, tidx, size)...);
+}
+
+
+template <typename KernelFunction, typename... KernelArgs>
+__global__ __launch_bounds__(default_block_size) void generic_kernel_2d(
+    size_type rows, size_type cols, KernelFunction fn, KernelArgs... args)
+{
+    auto tidx = thread::get_thread_id_flat();
+    auto col = tidx % cols;
+    auto row = tidx / cols;
+    if (row >= rows) {
+        return;
+    }
+    fn(row, col,
+       device_unpack_2d_impl<KernelArgs>::unpack(args, row, col, rows,
+                                                 cols)...);
+}
+
+
+}  // namespace hip
+}  // namespace kernels
+
+
+template <typename KernelFunction, typename... KernelArgs>
+void run_kernel(std::shared_ptr<const HipExecutor> exec, KernelFunction fn,
+                size_type size, KernelArgs &&... args)
+{
+    hip::device_guard guard{exec->get_device_id()};
+    constexpr auto block_size = kernels::hip::default_block_size;
+    auto num_blocks = ceildiv(size, block_size);
+    hipLaunchKernelGGL(kernels::hip::generic_kernel_1d, num_blocks, block_size,
```

Comment:
It's funny that `hipLaunchKernelGGL` itself is a higher-order function to launch other functions, but in your case it launches a function that "launches" yet another function.  😆